sound/soc/sof/intel/mtl.c

```diff
@@ -406,6 +406,41 @@ static int mtl_dsp_core_power_down(struct snd_sof_dev *sdev, int core)
 	return ret;
 }
 
+static int mtl_dsp_power_down_dsp_subsystem(struct snd_sof_dev *sdev)
+{
+	u32 dsphfdsscs;
+	u32 cpa;
+
+	/* Set the DSP subsystem power down */
+	snd_sof_dsp_update_bits(sdev, HDA_DSP_BAR, MTL_HFDSSCS,
+				MTL_HFDSSCS_SPA_MASK, 0);
+
+	/* Wait for unstable CPA read (1 then 0 then 1) just after setting SPA bit */
+	usleep_range(1000, 1010);
+
+	/* poll with timeout to check if operation successful */
+	cpa = MTL_HFDSSCS_CPA_MASK;
+	dsphfdsscs = snd_sof_dsp_read(sdev, HDA_DSP_BAR, MTL_HFDSSCS);
+	return snd_sof_dsp_read_poll_timeout(sdev, HDA_DSP_BAR, MTL_HFDSSCS, dsphfdsscs,
+					    (dsphfdsscs & cpa) == 0, HDA_DSP_REG_POLL_INTERVAL_US,
+					    HDA_DSP_RESET_TIMEOUT_US);
+}
+
+static int mtl_power_down_dsp(struct snd_sof_dev *sdev)
+{
+	int ret;
+
+	/* first power down core */
+	ret = mtl_dsp_core_power_down(sdev, SOF_DSP_PRIMARY_CORE);
+	if (ret) {
+		dev_err(sdev->dev, "mtl dsp power down error, %d\n", ret);
+		return ret;
+	}
+
+	/* power down dsp subsysem */
+	return mtl_dsp_power_down_dsp_subsystem(sdev);
```

Comment:
no need to have a separate function here and the names are confusing. Please just move the code inside this function.

Reply:
it's not so bad @ranj063, see below the helper is used in the mlt_suspend() routine

Reply:
@plbossart the suspend callback needs to change as well to use the whole power_down_dsp op. I have a patch coming up after this PR. Maybe I can change it then.

Reply:
makes sense, thanks for the precision. So let's avoid the split function @fredoh9